File: smoke_tests/php/mass_assignment.php
<?php
// Mass Assignment vulnerabilities in PHP

class User {
    public $id;
    public $username;
    public $email;
    public $password;
    public $is_admin = false;  // Sensitive!
    public $balance = 0;       // Sensitive!
    public $role = 'user';     // Sensitive!
}

// Test 1: Direct property assignment from $_POST
function create_user_unsafe() {
    $user = new User();
    // VULNERABLE: All $_POST values assigned
    foreach ($_POST as $key => $value) {
        if (property_exists($user, $key)) {
            $user->$key = $value;
        }
    }
    return $user;
}

// Test 2: Using extract()
function create_with_extract() {
    $user = new User();
    // VULNERABLE: extract() creates variables from array
    extract($_POST);
    $user->username = $username ?? '';
    $user->email = $email ?? '';
    $user->is_admin = $is_admin ?? false;  // Oops!
    return $user;
}

// Test 3: Laravel-style mass assignment without $fillable
class UnsafeModel {
    public $attributes = [];

    public function fill($data) {
        // VULNERABLE: No attribute filtering
        foreach ($data as $key => $value) {
            $this->attributes[$key] = $value;
        }
        return $this;
    }
}

function create_laravel_style() {
    $model = new UnsafeModel();
    // VULNERABLE: All request data
    return $model->fill($_POST);
}

// Test 4: __set magic method
class MagicUser {
    private $data = [];

    public function __set($name, $value) {
        // VULNERABLE: No filtering
        $this->data[$name] = $value;
    }

    public function __get($name) {
        return $this->data[$name] ?? null;
    }
}

function create_magic_user() {
    $user = new MagicUser();
    // VULNERABLE: Any property can be set
    foreach ($_POST as $key => $value) {
        $user->$key = $value;
    }
    return $user;
}

// Test 5: array_merge with user data
function merge_user_data() {
    $defaults = ['role' => 'user', 'is_admin' => false];
    $user_data = $_POST;
    // VULNERABLE: User data overrides defaults
    return array_merge($defaults, $user_data);
}

// Test 6: Doctrine entity without protection
class DoctrineUser {
    private $id;
    private $username;
    private $email;
    private $isAdmin;

    public function fromArray($data) {
        // VULNERABLE: No field filtering
        foreach ($data as $key => $value) {
            $setter = 'set' . ucfirst($key);
            if (method_exists($this, $setter)) {
                $this->$setter($value);
            }
        }
    }

    public function setIsAdmin($value) {
        $this->isAdmin = $value;
    }
}

// Test 7: PDO insert with all fields
function insert_user_pdo() {
    $pdo = get_pdo();
    $fields = array_keys($_POST);
    $placeholders = array_map(fn($f) => ":$f", $fields);

    // VULNERABLE: Inserting all POST fields
    $sql = "INSERT INTO users (" . implode(',', $fields) . ") VALUES (" . implode(',', $placeholders) . ")";
    $stmt = $pdo->prepare($sql);
    $stmt->execute($_POST);
}

// Test 8: JSON decode to object
function create_from_json() {
    $json = file_get_contents('php://input');
    // VULNERABLE: All JSON properties become object properties
    $user = json_decode($json);
    return $user;
}

// Test 9: Symfony form without field configuration
function handle_form() {
    // Pseudo-code for Symfony without proper form configuration
    $form = create_form(User::class);
    $form->submit($_POST);  // VULNERABLE: All fields submitted
    return $form->getData();
}

// Test 10: Yii2 model load
class Yii2User {
    public $username;
    public $email;
    public $is_admin;

    public function load($data) {
        // VULNERABLE: No safe attributes defined
        foreach ($data as $key => $value) {
            if (property_exists($this, $key)) {
                $this->$key = $value;
            }
        }
        return true;
    }
}

function get_pdo() {
    // Return PDO instance
    return new PDO('sqlite::memory:');
}

function create_form($class) {
    // Placeholder
    return new class {
        public function submit($data) {}
        public function getData() { return new User(); }
    };
}
?>

File: smoke_tests/csharp/mass_assignment.cs
// Mass Assignment vulnerabilities in C#
using System;
using System.Web.Mvc;

namespace VulnerableApp
{
    // Vulnerable model with sensitive properties
    public class User
    {
        public int Id { get; set; }
        public string Username { get; set; }
        public string Email { get; set; }
        public string Password { get; set; }
        public bool IsAdmin { get; set; }  // Sensitive!
        public decimal Balance { get; set; }  // Sensitive!
        public string Role { get; set; }  // Sensitive!
    }

    public class UserController : Controller
    {
        // Test 1: Direct model binding without protection
        [HttpPost]
        public ActionResult Create(User user)
        {
            // VULNERABLE: All properties bound including IsAdmin
            // Attacker can POST IsAdmin=true
            SaveUser(user);
            return Ok();
        }

        // Test 2: UpdateModel without whitelist
        [HttpPost]
        public ActionResult Update(int id)
        {
            var user = GetUserById(id);
            // VULNERABLE: Updates all posted properties
            UpdateModel(user);
            SaveUser(user);
            return Ok();
        }

        // Test 3: TryUpdateModel without include list
        [HttpPost]
        public ActionResult Edit(int id)
        {
            var user = GetUserById(id);
            // VULNERABLE: No property filtering
            if (TryUpdateModel(user))
            {
                SaveUser(user);
            }
            return Ok();
        }

        // Test 4: Bind attribute missing sensitive properties
        [HttpPost]
        public ActionResult Register([Bind(Include = "Username,Email,Password,IsAdmin")] User user)
        {
            // VULNERABLE: IsAdmin should not be in Include list
            SaveUser(user);
            return Ok();
        }

        // Test 5: FormCollection to model mapping
        [HttpPost]
        public ActionResult FromForm(FormCollection form)
        {
            var user = new User();
            // VULNERABLE: Manual binding without filtering
            user.Username = form["Username"];
            user.Email = form["Email"];
            user.IsAdmin = bool.Parse(form["IsAdmin"] ?? "false");  // Should not bind
            SaveUser(user);
            return Ok();
        }

        // Test 6: JSON deserialization mass assignment
        [HttpPost]
        public ActionResult JsonCreate()
        {
            var serializer = new System.Web.Script.Serialization.JavaScriptSerializer();
            string json = new System.IO.StreamReader(Request.InputStream).ReadToEnd();
            // VULNERABLE: All JSON properties deserialized
            var user = serializer.Deserialize<User>(json);
            SaveUser(user);
            return Ok();
        }

        // Test 7: ModelState with direct property access
        [HttpPost]
        public ActionResult Partial(int id)
        {
            var user = GetUserById(id);
            foreach (string key in Request.Form.AllKeys)
            {
                // VULNERABLE: Setting properties by name from form
                var prop = typeof(User).GetProperty(key);
                if (prop != null)
                {
                    prop.SetValue(user, Convert.ChangeType(Request.Form[key], prop.PropertyType));
                }
            }
            SaveUser(user);
            return Ok();
        }

        // Test 8: Automapper without configuration
        public ActionResult MapFromDto(UserDto dto)
        {
            var user = new User();
            // VULNERABLE: If automapper maps IsAdmin
            // AutoMapper.Mapper.Map(dto, user);
            foreach (var prop in typeof(UserDto).GetProperties())
            {
                var targetProp = typeof(User).GetProperty(prop.Name);
                if (targetProp != null)
                {
                    targetProp.SetValue(user, prop.GetValue(dto));
                }
            }
            SaveUser(user);
            return Ok();
        }

        // Placeholder methods
        private User GetUserById(int id) => new User { Id = id };
        private void SaveUser(User user) { }
    }

    public class UserDto
    {
        public string Username { get; set; }
        public string Email { get; set; }
        public bool IsAdmin { get; set; }  // Should not be in DTO
    }
}
